man/mkbinhmm.1
'\" t
.\"     Title: 
    mkbinhmm
  
.\"    Author: 
.\" Generator: DocBook XSL Stylesheets v1.76.1 <http://docbook.sf.net/>
.\"      Date: 12/19/2013
.\"    Manual: 
.\"    Source: 
.\"  Language: English
.\"
.TH "MKBINHMM" "1" "12/19/2013" ""
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"

    mkbinhmm
   \- convert HMM definition file in HTK ascii format to Julius binary format
.SH "SYNOPSIS"
.HP \w'\fBmkbinhmm\fR\ 'u
\fBmkbinhmm\fR [\-htkconf\ \fIHTKConfigFile\fR] {hmmdefs_file} {binhmm_file}
.SH "DESCRIPTION"
.PP

\fBmkbinhmm\fR
convert an HMM definition file in HTK ascii format into a binary HMM file for Julius\&. It will greatly speed up the launch process\&.
.PP
You can also embed acoustic analysis condition parameters needed for recognition into the output file\&. To embed the parameters, specify the HTK Config file you have used to extract acoustic features for training the HMM by the optione "\fB\-htkconf\fR"\&.
.PP
The embedded parameters in a binary HMM format will be loaded into Julius automatically, so you do not need to specify the acoustic feature options at run time\&. It will be convenient when you deliver an acoustic model\&.
.PP
You can also specify binary file as the input\&. This can be used to update the old binary format into new one, or to embed the config parameters into the already existing binary files\&. If the input binhmm already has acoustic analysis parameters embedded, they will be overridden by the specified values\&.
.PP

\fBmkbinhmm\fR
can read gzipped file as input\&.
.SH "OPTIONS"
.PP
\fB \-htkconf \fR \fIHTKConfigFile\fR
.RS 4
HTK Config file you used at training time\&. If specified, the values are embedded to the output file\&.
.RE
.PP
\fIhmmdefs_file\fR
.RS 4
The source HMm definitino file in HTK ascii format or Julius binary format\&.
.RE
.PP
\fIhmmdefs_file\fR
.RS 4
Output file\&.
.RE
.SH "EXAMPLES"
.PP
Convert HTK ascii format HMM definition file into Julius binary file:
.sp .if n \{\ .RS 4 .\} .nf % \fBmkbinhmm\fR hmmdefs\&.ascii binhmm .fi .if n \{\ .RE .\}
Furthermore, embed acoustic feature parameters as specified by Config file
.sp .if n \{\ .RS 4 .\} .nf % \fBmkbinhmm\fR \-htkconf Config hmmdefs\&.ascii binhmm .fi .if n \{\ .RE .\}
Embed the acoustic parameters into an existing binary file
.sp .if n \{\ .RS 4 .\} .nf % \fBmkbingram\fR \-htkconf Config old_binhmm new_binhmm .fi .if n \{\ .RE .\}
.SH "SEE ALSO"
.PP

\fB julius \fR( 1 )
,
\fB mkbingram \fR( 1 )
,
\fB mkbinhmmlist \fR( 1 )
.SH "COPYRIGHT"
.PP
Copyright (c) 1991\-2013 Kawahara Lab\&., Kyoto University
.PP
Copyright (c) 1997\-2000 Information\-technology Promotion Agency, Japan
.PP
Copyright (c) 2000\-2005 Shikano Lab\&., Nara Institute of Science and Technology
.PP
Copyright (c) 2005\-2013 Julius project team, Nagoya Institute of Technology
.SH "LICENSE"
.PP
The same as Julius\&.


ChangeLog
2008-05-17 13:06  sumomo

	* libjulius/src/: instance.c, m_fusion.c, m_options.c: Now GMM
	  module uses the analysis parameter of the given AM when "-AM_GMM"
	  is not explicitly specified.	Now output error when GMM has
	  different sampling rate from AM.

2008-05-13 15:32  sumomo

	* jcontrol/jcontrol.c: Flush output after each message.

2008-05-13 15:32  sumomo

	* libsent/src/util/readfile.c: Fixed file name prompt to ignore
	  empty input correctly.

2008-05-09 15:00  sumomo

	* libjulius/src/m_usage.c: Grouped the usage messages by the
	  belonging modules.

2008-05-09 14:59  sumomo

	* libjulius/src/m_options.c: New option "-usepower" to use power
	  instead of magnitude in filterbank analysis.

2008-05-09 14:58  sumomo

	* libsent/: include/sent/mfcc.h, src/hmminfo/read_binhmm.c,
	  src/hmminfo/write_binhmm.c, src/wav2mfcc/mfcc-core.c: Added
	  support for USEPOWER in filterbank analysis.

2008-05-09 14:57  sumomo

	* libsent/src/wav2mfcc/para.c: Add support for USEPOWER.  Fixed
	  -htkconf to assume SOURCERATE=625 (16kHz) if not exist.

2008-05-08 14:45  sumomo

	* libjulius/: include/julius/extern.h, src/adin-cut.c,
	  src/m_adin.c: Add tiny value check for frequency.

2008-05-08 14:44  sumomo

	* libsent/src/adin/adin_mic_linux_alsa.c: Fixed failure to startup
	  when using multiple sound cards on ALSA.

2008-05-08 14:43  sumomo

	* libsent/src/adin/adin_file.c: Fixed segfault when failed to open
	  audio file.

2008-04-27 22:06  sumomo

	* libjulius/: include/julius/jconf.h, src/default.c, src/m_info.c,
	  src/m_options.c, src/m_usage.c, src/search_bestfirst_main.c: Add
	  an option "-fallback1pass" to use the 1st pass result as final
	  result when the 2nd pass fails.  This was the default on 3.x.

2008-04-27 22:06  sumomo

	* libjulius/src/version.c.in: Update copyright output.

2008-03-28 13:52  sumomo

	* libsent/src/adin/adin_mic_linux_alsa.c: Fix ALSA A/D-in to use
	  device other than "default".

2008-03-28 12:57  sumomo

	* julius/record.c: Fixed record sometimes fails.

2008-03-28 12:51  sumomo

	* libjulius/src/adin-cut.c: Moved CALLBACK_ADIN_TRIGGERED to adin
	  thread.

2008-03-25 17:12  sumomo

	* libjulius/src/pass1.c: Fixed GMM rejection not working well on
	  microphone input (enbuged on 4.0.1).

2008-03-21 11:59  sumomo

	* libsent/src/wav2mfcc/para.c: Add warning when HTK Config given by
	  "-htkconf" has unsupported values.

2008-03-20 18:08  sumomo

	* libsent/: include/sent/ngram2.h, include/sent/util.h,
	  src/ngram/ngram_access.c, src/ngram/ngram_compact_context.c,
	  src/ngram/ngram_read_arpa.c, src/ngram/ngram_read_bin.c,
	  src/ngram/ngram_util.c, src/ngram/ngram_write_bin.c,
	  src/util/endian.c: Update N-gram codes to (hopefully) allow 64bit
	  index in the future.

2008-03-20 18:06  sumomo

	* julius/: charconv_win32.c, libjcode/jlib.h, libjcode/libjcode.c:
	  Fixed euc-jp conversion not working on some string.

2008-03-18 02:00  sumomo

	* libjulius/src/: adin-cut.c, recogmain.c: Fixed
	  CALLBACK_EVENT_SPEECH_* call timing when -spsegment/VAD enabled.

2008-03-18 01:59  sumomo

	* libjulius/src/version.c.in, libsent/src/util/confout.c: Fix
	  typos.

2008-03-18 01:58  sumomo

	* julius/main.c, julius/output_stdout.c,
	  libjulius/include/julius/global.h, libjulius/src/callback.c,
	  libjulius/src/m_options.c, libjulius/src/m_usage.c:
	  "-callbackdebug" moved from Julius to JuliusLib.

2008-03-14 22:28  sumomo

	* libsent/src/ngram/ngram_compact_context.c: Fixed a bug that
	  prevents reading big N-gram.

2008-03-12 15:43  tag release_4_0_1

2008-03-12 15:43  sumomo

	* 00readme-ja.txt, 00readme.txt, ChangeLog, LICENSE.txt,
	  Release-ja.txt, Release.txt, libjulius/configure,
	  libjulius/configure.in, libsent/configure, libsent/configure.in,
	  support/build-all.sh: Release of 4.0.1.

2008-03-12 14:48  sumomo

	* libjulius/src/pass1.c: Fixed compilation failure enbuged at last
	  fix...

2008-03-10 11:39  sumomo

	* libjulius/src/: pass1.c, realtime-1stpass.c, recogmain.c: Fix CMN
	  not working well on spsegment/GMM-VAD/decoder-VAD mode.

2008-03-02 02:10  sumomo

	* libsent/: configure, configure.in, include/sent/config.h.in,
	  src/adin/adin_mic_linux_alsa.c: Fixed ALSA recording to work.
	  Many codes are updated.

2008-02-21 13:10  sumomo

	* libjulius/src/beam.c: Word recognition results are not correctly
	  sorted when using "-output N".

2008-02-18 23:35  sumomo

	* julius/module.c: Added a module command "GRAMINFO" to return
	  grammar info.

2008-02-18 23:32  sumomo

	* libjulius/src/m_fusion.c: Now -spsegment detects error when pause
	  word is not defined in dictionary.

2008-02-15 03:26  sumomo

	* libsent/src/phmm/mkwhmm.c: Fixed bug of immediate stop when no sp
	  model in multi-path mode.

2008-02-15 03:25  sumomo

	* libjulius/src/: m_info.c, search_bestfirst_v1.c,
	  search_bestfirst_v2.c, word_align.c: Fixed hungup when sp model
	  not exist on AM in multipath.

2008-02-14 13:44  sumomo

	* libjulius/: include/julius/extern.h, src/multi-gram.c: Changes
	  for better grammar handling: - multigram_add() returns the newly
	  assigned grammar id.	- a function is added to get grammar id
	  from name string.  - modified STAT message to output grammar name
	  together with id.

2008-02-14 13:41  sumomo

	* julius/module.c: Extended ACTIVATEGRAM, DEACTIVATEGRAM, DELGRAM
	  to specify grammar by name.

2008-02-13 17:49  sumomo

	* libjulius/include/julius/extern.h: Updated for some function
	  changes.

2008-02-13 17:48  sumomo

	* libjulius/src/: outprob_style.c, search_bestfirst_v1.c,
	  search_bestfirst_v2.c: Global variables and work area has been
	  moved to RecogProcess and StackDecode.

2008-02-13 17:47  sumomo

	* libjulius/src/wchmm.c: qsort function switched to qsort_reentrant
	  in libsent for modularity.

2008-02-13 17:46  sumomo

	* libjulius/src/confnet.c: Updated for being re-entrant.

2008-02-13 17:45  sumomo

	* libjulius/src/: m_jconf.c, m_options.c, ngram_decode.c,

mkbinhmm/mkbinhmm.c
/*
 * Copyright (c) 2003-2013 Kawahara Lab., Kyoto University 
 * Copyright (c) 2003-2005 Shikano Lab., Nara Institute of Science and Technology
 * Copyright (c) 2005-2013 Julius project team, Nagoya Institute of Technology
 * All rights reserved
 */

/* mkbinhmm --- read in ascii hmmdefs file and write in binary format */

/* $Id: mkbinhmm.c,v 1.7 2013/06/20 17:14:27 sumomo Exp $ */

#include <sent/stddefs.h>
#include <sent/htk_hmm.h>


HTK_HMM_INFO *hmminfo;
Value para, para_htk;


static void
usage(char *s)
{
  printf("mkbinhmm: convert HMM definition file to binary format for Julius\n");
  printf("usage: %s [-htkconf HTKConfig] hmmdefs binhmm\n", s);
  printf("\nLibrary configuration: ");
  confout_version(stdout);
  confout_am(stdout);
  printf("\n");
}


int
main(int argc, char *argv[])
{
  FILE *fp;
  char *infile;
  char *outfile;
  char *conffile;
  int i;

  infile = outfile = conffile = NULL;
  for(i=1;i<argc;i++) {
    if (strmatch(argv[i], "-C") || strmatch(argv[i], "-htkconf")) {
      if (++i >= argc) {
	usage(argv[0]);
	return -1;
      }
      conffile = argv[i];
    } else {
      if (infile == NULL) {
	infile = argv[i];
      } else if (outfile == NULL) {
	outfile = argv[i];
      } else {
	usage(argv[0]);
	return -1;
      }
    }
  }
  if (infile == NULL || outfile == NULL) {
    usage(argv[0]);
    return -1;
  }

  hmminfo = hmminfo_new();

  printf("---- reading hmmdefs ----\n");
  printf("filename: %s\n", infile);

  /* read hmmdef file */
  undef_para(&para);
  if (init_hmminfo(hmminfo, infile, NULL, &para) == FALSE) {
    fprintf(stderr, "--- terminated\n");
    return -1;
  }

  if (conffile != NULL) {
    /* if input HMMDEFS already has embedded parameter
       they will be overridden by the parameters in the config file */
    printf("\n---- reading HTK Config ----\n");
    if (para.loaded == 1) {
      printf("Warning: input hmmdefs has acoustic analysis parameter information\n");
      printf("Warning: they are overridden by the HTK Config file...\n");
    }
    /* load HTK config file */
    undef_para(&para);
    if (htk_config_file_parse(conffile, &para) == FALSE) {
      fprintf(stderr, "Error: failed to read %s\n", conffile);
      return(-1);
    }
    /* set some parameters from HTK HMM header information */
    printf("\nsetting TARGETKIND and NUMCEPS from HMM definition header...");
    calc_para_from_header(&para, hmminfo->opt.param_type, hmminfo->opt.vec_size);
    printf("done\n");
    /* fulfill unspecified values with HTK defaults */
    printf("fulfill unspecified values with HTK defaults...");
    undef_para(&para_htk);
    make_default_para_htk(&para_htk);
    apply_para(&para, &para_htk);
    printf("done\n");
  }

  printf("\n------------------------------------------------------------\n");
  print_hmmdef_info(stdout, hmminfo);
  printf("\n");

  if (para.loaded == 1) {
    put_para(stdout, &para);
  }
  printf("------------------------------------------------------------\n");

  printf("---- writing ----\n");
  printf("filename: %s\n", outfile);
  
  if ((fp = fopen_writefile(outfile)) == NULL) {
    fprintf(stderr, "failed to open %s for writing\n", outfile);
    return -1;
  }
  if (write_binhmm(fp, hmminfo, (para.loaded == 1) ? &para : NULL) == FALSE) {
    fprintf(stderr, "failed to write to %s\n", outfile);
    return -1;
  }
  if (fclose_writefile(fp) != 0) {
    fprintf(stderr, "failed to close %s\n", outfile);
    return -1;
  }

  printf("\n");
  if (para.loaded == 1) {
    printf("binary HMM written to \"%s\", with acoustic parameters embedded for Julius.\n", outfile);
  } else {
    printf("binary HMM written to \"%s\"\n", outfile);
  }

  return 0;
}
